### PR TITLE
add API definition in Open API format and swagger runtime 

### PR DESCRIPTION
**Pull request by akihikokuroda**
# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/dashboard/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

**Comment by akihikokuroda** [comment]
Swagger doesn't support the same path with different operation.  So I put fake path for 
credentials put, credentials delete and pipelinerun put APIs

**Comment by AlanGreene** [comment]
> Swagger doesn't support the same path with different operation.

Can you clarify what you mean by this? If you mean having GET + POST on the same path for example this **is** supported.

We should not add 'fake' paths as that defeats the purpose of having this documentation.

**Comment by akihikokuroda** [comment]
It may be my misunderstanding.  I’ll try some after I come back from my vacation.

**Comment by akihikokuroda** [comment]
It was my misunderstanding.  Everything is fine now.

**Comment by a-roberts** [comment]
@akihikokuroda this is looking good! 

Is this something we should mention in our DEVELOPMENT.md docs too incase someone changes any of our APIs? 

Do they just run the Docker container again and then commit the generated json? 

In addition, can you mention in DEVELOPMENT.md how this could be maintained by anyone - why do we need another service for example?

**Comment by akihikokuroda** [comment]
I added statements in the DEVELOPMENT.md.
When the API is changed  tekton-dashboard-api.yaml msut be updated and the docker image must be rebuilt.

File: swagger/tekton-dashboard-api.yaml [hunk 1]
```diff
@@ -0,0 +1,788 @@
...
+        default: localhost
+      port:
+        default: '9097'
+      namespace:
+        default: tekton-pipelines
+        description: >-
+          namespace where the Tekton resources (Pipelines, Tasks, etc.) are
+          installed
+paths:
+  /pipeline/:
```
Comment (AlanGreene) [inline swagger/tekton-dashboard-api.yaml]: https://github.com/tektoncd/dashboard/pull/155 updated the APIs to use plurals, so for example here `pipeline` should be `pipelines`

File: DEVELOPMENT.md [hunk 2]
```diff
@@ -198,6 +198,8 @@ Run `npm run storybook:build` to build the static storybook files. The build art
 
 ## API definitions
 
+API definitions are documented in the Open API format in addition to the document below.  Please maintain the swagger/tekton-dashboard-api.yaml as well as this file when the API difinitions are changed.  The swagger UI  is available as the extension of the tekton dashboard.  Refer to swagger/README.md to enable it.
```
Comment (a-roberts) [inline DEVELOPMENT.md]: - Typo here (`definitions`), and some dubious whitespace
- `Tekton dashboard` (casing)
- Let's use `code` markdown too when referencing files
- `OpenAPI` not `Open API`, and we should add a link to https://swagger.io/docs/specification/about/

File: swagger/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+FROM swaggerapi/swagger-ui
+ENV SWAGGER_JSON=/foo/tekton-dashboard-api.yaml
+COPY tekton-dashboard-api.yaml /foo/
+
```
Comment (a-roberts) [inline swagger/Dockerfile]: Let's remove the whitespace, is /foo/ legit?

File: swagger/README.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# Tekton dashboard REST API - Open API
+
+The API definitions are in tekton-dashboard-api.yaml file
```
Comment (a-roberts) [inline swagger/README.md]: We can include a hyperlink in the readme to the actual file reference

File: DEVELOPMENT.md (hunk 2 above)
Comment (akihikokuroda) [reply DEVELOPMENT.md]: I fixed all

File: swagger/Dockerfile (hunk 3 above)
Comment (akihikokuroda) [reply swagger/Dockerfile]: Yes, the /foo is legit for the base swagger-ui image

File: swagger/README.md (hunk 4 above)
Comment (akihikokuroda) [reply swagger/README.md]: I added it.

File: swagger/tekton-dashboard-api.yaml (hunk 1 above)
Comment (akihikokuroda) [reply swagger/tekton-dashboard-api.yaml]: I updated them.

**Comment by AlanGreene** [comment]
Closing this old PR since we've removed most of our custom REST APIs.

**Pull request closed without merging**